### DISCUSSION TITLE
Space of `HingeConstraint` lambdas

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#629] [Q&A] [answered]
I'm not sure if this question even makes any sense, but I'm trying to make sense of this 2D vector that you get back from `HingeConstraint::GetTotalLambdaRotation` and I can't quite figure it out.

When dealing with something like `FixedConstraint` and the lambdas you get back from its `RotationEulerConstraintPart` things seem simple enough, where everything seems to be in world-space. I figured then that maybe the rotation lambdas for `HingeConstraint` just had its axes split up between the rotation around the hinge axis (`GetTotalLambdaRotationLimits`) and the other two (`GetTotalLambdaRotation`), but from my (primitive) testing this doesn't seem to hold true.

As an example, you can end up with roughly the same lambdas when rotating the hinge in a particular direction around a particular local-space axis, but positioning the whole setup in different world-space axes, leading me to believe the lambdas are in some kind of local-space, but then other times you don't.

Does it even make sense to try and treat the individual lambdas as specific axes for this particular constraint? If so, how would I go about transforming them to world-space? If not, does it make more sense to just grab the `.Length()` of all these different lambda vectors and expose them as a single float instead?

I'd be happy to make some type of repro if needed.

**Comment by jrouwe** [comment, marked as answer]
They are indeed in some constraint space:

https://github.com/jrouwe/JoltPhysics/blob/b1cd84951022a565a23f33ff77ffee9e71487bbc/Jolt/Physics/Constraints/ConstraintPart/HingeRotationConstraintPart.h#L96-L97

In general the first lambda is relative to (body 2's world space hinge axis).GetNormalizedPerpendicular(), the second relative to an axis perpendicular to the hinge and lambda 1's axis.

Unfortunately, if the hinge is rotated enough under an angle that's not the hinge axis, the axis changes (see a few lines higher) and after the solve step the rotation may have changed enough so that it becomes impossible to get the axis back (because GetNormalizedPerpendicular has discontinuities).

Maybe just exposing Length() is indeed the easiest solution.

**Reply by mihe** [reply]
Thank you for the speedy reply, as always.

`Length()` it is then. That seems to be how PhysX deals with its [break force/torque](https://nvidia-omniverse.github.io/PhysX/physx/5.2.1/_build/physx/latest/class_px_fixed_joint.html#_CPPv4N12PxFixedJoint13setBreakForceE6PxReal6PxReal) as well.
